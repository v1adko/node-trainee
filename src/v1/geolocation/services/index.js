export { default as geolocationService } from './geolocation-service';
export { default as coordinatesService } from './coordinates-service';
