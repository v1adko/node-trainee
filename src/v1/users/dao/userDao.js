import BaseDao from '../../../utils/baseDaoMongoose';
import User from '../models/user';

class UserDao extends BaseDao {
  create(username, password) {
    const user = new this.Model();
    user.username = username;
    user.password = password;
    return super.create(user);
  }
}

export default new UserDao(User);
