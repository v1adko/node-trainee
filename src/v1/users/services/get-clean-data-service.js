const getCleanData = (filterData) => {
  const cleanFilterData = {};
  Object.keys(filterData)
    .filter(key => filterData[key])
    .forEach((key) => {
      cleanFilterData[key] = filterData[key];
    });
  return cleanFilterData;
};

export default getCleanData;
