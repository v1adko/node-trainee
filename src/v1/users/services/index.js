export { default as passwordService } from './password-service';
export { default as modelService } from './model-service';
export { default as getCleanDataService } from './get-clean-data-service';
