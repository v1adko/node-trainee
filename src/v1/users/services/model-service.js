class ModelService {
  getSafeItem = (item) => {
    console.log(item);
    if (!item.safeFields) return {};

    const safeItem = {};

    item.safeFields.forEach((field) => {
      safeItem[field] = item[field];
    });

    return safeItem;
  };

  mapSafeItems = (key, items) => {
    if (!items) return {};

    const itemsMap = {};
    items.forEach((item) => {
      itemsMap[item[key]] = this.getSafeItem(item, item.safeFields);
    });
    return itemsMap;
  };
}

export default new ModelService();
