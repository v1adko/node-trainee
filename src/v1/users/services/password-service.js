import { compareSync, genSaltSync, hashSync } from 'bcryptjs';
import { RequestValidationError } from '../../../lib/errors';

class PasswordService {
  valid = (user, password) => compareSync(password, user.hash);

  change = (user, password, newPassword) => {
    if (!this.valid(user, password)) {
      throw new RequestValidationError(
        'Password is wrong, please check input data.'
      );
    }

    const changedUser = user;
    changedUser.password = newPassword;
  };

  generateSaltAndHash = (password) => {
    const saltAndHash = {};
    saltAndHash.salt = genSaltSync(10);
    saltAndHash.hash = hashSync(password, this.salt);
    return saltAndHash;
  };
}

export default new PasswordService();
