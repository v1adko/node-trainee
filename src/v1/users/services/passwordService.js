import { compareSync } from 'bcryptjs';

class PasswordService {
  valid = (user, password) => compareSync(password, user.hash);

  change = (user, password, newPassword) => {
    if (this.valid(user, password)) {
      const changedUser = user;
      changedUser.password = newPassword;
      return changedUser;
    }
    throw new Error('Password is wrong');
  };
}

export default new PasswordService();
