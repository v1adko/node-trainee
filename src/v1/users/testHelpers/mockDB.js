import userDao from '../dao';
import userMocks from './userMocks';
import permissions from '../../../constants/permissions';

const { getUsersFields, user: userFields, admin: adminFields } = userMocks;

class MockDB {
  constructor(DAO) {
    this.DAO = DAO;
  }

  createDefaultUsers = async (userCount = 3) => {
    const users = [];
    getUsersFields(userCount).forEach(async (element) => {
      const { username, password, role } = element;
      const user = await this.createUser(username, password, role);
      users.push(user);
    });
  };

  createDefaultUser = async () => {
    const { username, password, role } = userFields;
    const user = await this.createUser(username, password, role);
    return user;
  };

  createDefaultAdmin = async () => {
    const { username, password, role } = adminFields;
    const user = await this.createUser(username, password, role);
    return user;
  };

  createUser = async (username, password, role = permissions.USER.value) => {
    const createdUser = this.DAO.create(username, password, role);
    return createdUser;
  };

  getAll = async () => {
    const users = await this.DAO.getAll();
    return users;
  };

  cleanDB = async () => {
    await this.DAO.Model.remove();
  };
}

export default new MockDB(userDao);
