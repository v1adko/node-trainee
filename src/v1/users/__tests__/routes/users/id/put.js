import simulate from '../../../../../../utils/tests-utils/request-helper';
import jwtService from '../../../../../services/jwt-service';
import User from '../../../../user-model';
import userDao from '../../../../user-dao';
import mockDB from '../../../../../../utils/tests-utils/mock-db';
import permissions from '../../../../../../constants/permissions';
import { passwordService } from '../../../../services/index';
import UserFields from '../../../../../../utils/tests-utils/test-user-fields';

const ROUTE = '/v1/users';

const {
  username,
  password,
  newUsername,
  newPassword,
  invalidUserId,
  invalidToken,
  shortUsername,
  invalidRole,
  notExistingUserId
} = UserFields;
const newRole = permissions.ADMIN.value;
let user;
let userToken;
let adminToken;

async function clean() {
  await mockDB.cleanDB();
  userToken = null;
}

async function createUserAndAccessTokens() {
  const userActor = new User();
  const adminActor = new User();
  adminActor.role = permissions.ADMIN.value;
  userToken = jwtService.generateJwt(userActor);
  adminToken = jwtService.generateJwt(adminActor);
  user = await mockDB.createUser(username, password);
}

beforeAll(clean);
afterAll(mockDB.closeConnection);

describe.skip(`Test the ${ROUTE}/:id path`, () => {
  afterEach(clean);
  beforeEach(createUserAndAccessTokens);

  it('should change user by id', async () => {
    const route = `${ROUTE}/${user.id}`;
    const body = {
      username: newUsername,
      password: newPassword,
      role: newRole
    };
    const { message } = await simulate.put(route, 200, body, adminToken);
    const changedUser = await userDao.getById(user.id);
    const { id, username: name, role } = changedUser;
    const passwordChecked = passwordService.valid(changedUser, newPassword);

    expect(message).toMatchSnapshot();
    expect(id).toEqual(user.id);
    expect(name).toBe(newUsername);
    expect(role).toEqual(newRole);
    expect(passwordChecked).toBe(true);
  });

  it('should change only username', async () => {
    const route = `${ROUTE}/${user.id}`;
    const body = { username: newUsername };
    const { message } = await simulate.put(route, 200, body, adminToken);
    const changedUser = await userDao.getById(user.id);
    const { id, username: name, role } = changedUser;
    const passwordChecked = passwordService.valid(changedUser, password);

    expect(message).toMatchSnapshot();
    expect(id).toEqual(user.id);
    expect(name).toBe(newUsername);
    expect(role).toEqual(user.role);
    expect(passwordChecked).toBe(true);
  });

  it('should change only password', async () => {
    const route = `${ROUTE}/${user.id}`;
    const body = { password: newPassword };
    const { message } = await simulate.put(route, 200, body, adminToken);
    const changedUser = await userDao.getById(user.id);
    const { id, username: name, role } = changedUser;
    const passwordChecked = passwordService.valid(changedUser, newPassword);

    expect(message).toMatchSnapshot();
    expect(id).toEqual(user.id);
    expect(name).toBe(user.username);
    expect(role).toEqual(user.role);
    expect(passwordChecked).toBe(true);
  });

  it('should not return user in response on GET method, because user token is not valid', async () => {
    const route = `${ROUTE}/${user.id}`;
    const body = {
      username: newUsername,
      password: newPassword,
      role: newRole
    };
    const { auth, message } = await simulate.put(
      route,
      401,
      body,
      invalidToken
    );

    expect(auth).toBe(false);
    expect(message).toMatchSnapshot();
  });

  it('should not return user in response on GET method, because user token is not have enough permissions', async () => {
    const route = `${ROUTE}/${user.id}`;
    const body = {
      username: newUsername,
      password: newPassword,
      role: newRole
    };
    const { message } = await simulate.put(route, 403, body, userToken);

    expect(message).toMatchSnapshot();
  });

  it('should not return user in response on GET method, because user id is wrong', async () => {
    const route = `${ROUTE}/${invalidUserId}`;
    const body = {
      username: newUsername,
      password: newPassword,
      role: newRole
    };
    const { message } = await simulate.put(route, 400, body, adminToken);

    expect(message).toMatchSnapshot();
  });

  it('should not return user in response on GET method, because user is not exist', async () => {
    const route = `${ROUTE}/${notExistingUserId}`;
    const body = {
      username: newUsername,
      password: newPassword,
      role: newRole
    };
    const { message } = await simulate.put(route, 500, body, adminToken);

    expect(message).toMatchSnapshot(); // TODO: Fix it, when will do errors
  });

  it('should not update user because new username less than 6 symbols', async () => {
    const route = `${ROUTE}/${user.id}`;
    const body = { username: shortUsername };
    const { message } = await simulate.put(route, 400, body, adminToken);

    expect(message).toMatchSnapshot();
  });

  it('should not update user because role is invalid', async () => {
    const route = `${ROUTE}/${user.id}`;
    const body = { role: invalidRole };
    const { message } = await simulate.put(route, 400, body, adminToken);

    expect(message).toMatchSnapshot();
  });
});
