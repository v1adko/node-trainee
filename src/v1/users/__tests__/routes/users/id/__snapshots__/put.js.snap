// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the /v1/users/:id path should change only password 1`] = `"User was updated"`;

exports[`Test the /v1/users/:id path should change only username 1`] = `"User was updated"`;

exports[`Test the /v1/users/:id path should change user by id 1`] = `"User was updated"`;

exports[`Test the /v1/users/:id path should not return user in response on GET method, because user id is wrong 1`] = `"\\"id\\" must only contain hexadecimal characters"`;

exports[`Test the /v1/users/:id path should not return user in response on GET method, because user is not exist 1`] = `"Cannot set property 'username' of null"`;

exports[`Test the /v1/users/:id path should not return user in response on GET method, because user token is not have enough permissions 1`] = `"Access was denied. Not enough permissions."`;

exports[`Test the /v1/users/:id path should not return user in response on GET method, because user token is not valid 1`] = `"Invalid token, please repeat authentication."`;

exports[`Test the /v1/users/:id path should not update user because new username less than 6 symbols 1`] = `"\\"username\\" length must be at least 6 characters long"`;

exports[`Test the /v1/users/:id path should not update user because role is invalid 1`] = `"\\"role\\" must be one of [user, moderator, admin]"`;
